ProjFS.Mac/PrjFSKext/Memory.hpp

```diff
@@ -18,21 +18,27 @@ void Memory_Free(void* buffer, uint32_t size);
 template <typename T>
 T* Memory_AllocArray(uint32_t arrayLength)
 {
-    size_t allocBytes = arrayLength * sizeof(T);
-    if (allocBytes > UINT32_MAX)
+    uint32_t allocBytes;
+    if (__builtin_umul_overflow(arrayLength, sizeof(T), &allocBytes))
     {
+        // Overflow occurred.
         return nullptr;
     }
-    
-    return static_cast<T*>(Memory_Alloc(static_cast<uint32_t>(allocBytes)));
+
+    return static_cast<T*>(Memory_Alloc(allocBytes));
 }
 
 template <typename T>
-void Memory_FreeArray(T* array, uint32_t arrayLength)
+bool Memory_FreeArray(T* array, uint32_t arrayLength)
```

Comment:
Can we leave this as return type `void`?

`Memory_FreeArray` should only be called for memory that was allocated with `Memory_AllocArray`, and `Memory_AllocArray` already validated that there was no overflow when allocating.

What about something like:

```
void Memory_FreeArray(T* array, uint32_t arrayLength)
{
    uint32_t arrayBytes;
    if (__builtin_umul_overflow(arrayLength, sizeof(T), &arrayBytes)) 
    {
        // Overflow should never occur as Memory_AllocArray should have prevented it
        assert(false);
    }
    else
    {
        Memory_Free(array, arrayBytes);
    }
}
```

Reply:
That's a good point. I've incorporated that feedback.